### ISSUE TITLE
Difference in spectrogram implementation compared to the paper - potential degradation

### ISSUE THREAD

**Issue by shenberg** [google/visqol#57]
Hi, I've read the papers describing ViSQOL, and tried to wrap my head around the implementation you provided.

The second paper, "Objective Assessment of Perceptual Audio
Quality Using ViSQOLAudio", describes the spectrogram procedure (Part III.C) as:

"a short-time Fourier transform is performed with a 32 band Gammatone filter bank with a minimum frequency of..."

And in the following paper, "ViSQOL v3", a change is described to use an 80ms window with 25% overlap.

When examining the code, I saw that the Gammatone filterbank is implemented directly in the time domain. However, I did notice that a windowing function was not applied (the paper describes a hamming window, some comments in the code still refer to it). Won't this cause windowing artifacts which degrade the spectrogram? Is the change to an 80ms window enough to counteract this?

**Comment by mchinen** [comment]
Great question.  This change was implemented before I was involved, so I may be missing some context.  However, this is my understanding.  The gamamtone filterbank is used directly on the time domain signal instead of FFT. The gammatone signal is inherently windowed due to the shape of the gamma distribution, so unlike a regular STFT, windowing is not necessary.  It may be good to confirm this with the original authors.  See the image here for more info on the gammatone filterbank.
https://en.wikipedia.org/wiki/Gammatone_filter

**Comment by shenberg** [comment]
I did some investigation into the matter.
 I found a Python implementation of the same gammatone filter-bank, over at [https://github.com/detly/gammatone/blob/master/gammatone/filters.py] and used it to create a python mock of the implementation in visqol, as best I understand it (main logic in `gammatone_spectrogram.c`). It looks bad.
 
 ```python
import librosa
import librosa.display
import numpy as np
from gammatone.filters import centre_freqs, make_erb_filters

# from src/visqol_manager.cc
kNumBandsAudio = 32
kMinimumFreq = 50
kOverlap = 0.25
# from src/include/analysis_window.h
window_duration = 0.08

# compare src/equivalent_rectangular_bandwidth.cc to function make_erb_filters()
# they use the same constants and have similar logic, and seem based on 
# code written by Malcolm Slaney on June 11, 1998, including dividing the order-8 filter into 4 order-2 sections

fs = 48000
window_duration_samples = int(fs * window_duration)
hop_length_samples = int(fs * window_duration*kOverlap)

# python logic uses max frequency of fs / 2, exactly like in src/gammatone_spectrogram_builder.cc line 41
erb_frequencies = centre_freqs(fs, kNumBandsAudio, kMinimumFreq)
gammatone_filters = make_erb_filters(fs, erb_frequencies)


audio, _ = librosa.load(librosa.example('trumpet', hq=True), sr=fs)
# let's look at only 1 second
audio = audio[:fs]

windowed_audio = librosa.util.frame(audio, window_duration_samples, hop_length_samples)
# shape of windowed_audio is now (3840, 47) - 47 windows of 80ms overlapping by 75%

# for comparison's sake - we'll also use a window function
window = np.hanning(window_duration_samples)

output = np.zeros((filters.shape[0], windowed_audio.shape[1]), dtype=np.float32)
output_windowed = np.zeros((filters.shape[0], windowed_audio.shape[1]), dtype=np.float32)
for i in tqdm(range(windowed_audio.shape[1])):
    # calculate with no windowing
    filtered = erb_filterbank(windowed_audio[:, i], filters)
    output[:, i] = np.sum(filtered ** 2, axis=1)
    # calculate with windowing
    filtered = erb_filterbank(windowed_audio[:, i] * window, filters)
    output_windowed[:, i] = np.sum(filtered ** 2, axis=1)

# display log-power spectrograms, flip rows of outputs so low frequencies are at the bottom
fig, axes = plt.subplots(2,1, figsize=(8,8))
axes[0].set_title('No Window Function')
librosa.display.specshow(
    librosa.power_to_db(output[::-1], ref=np.max),
    ax=axes[0]
)
axes[1].set_title('Hanning Window')
librosa.display.specshow(
    librosa.power_to_db(output_windowed[::-1], ref=np.max),
    ax=axes[1]
)
plt.show()
```
The output as I run it on my system looks like this:
![image](https://user-images.githubusercontent.com/653972/140792793-ccb47745-11d6-4829-a423-2466e5f8bebf.png)
As far as I can tell, there are severe windowing artifacts in the gammatone spectrogram. A next step would be to check if this happens with the actual C++ code in VISQOL, though I see no reason for the results to be different. I hope this helps some.

**Comment by mchinen** [comment]
That is a very nice analysis, thank you for that.  I agree that we should look into this further.  looked in the trumpet file and a log-frequency STFT does not have the vertical banding artifacts. With this evidence I'm inclined to think that the artifacts are in ViSQOL as well.

**Comment by mchinen** [comment]
An update: I've implemented the Hann window and tested on retrained models on our internal ViSQOL.  For speech, there were some improvements in MSE (.57 to .50 for exponential model, .15 to .15 (unchanged) for lattice models) .  For audio SVR models, the metrics did not change significantly for most audio samples (e.g. typically less than .05 MOS), but there are a few samples such as low bitrate music where the subjective score is low that have a significant change in estimated MOS. 
 Since this requires retraining speech models, I'll be pushing this out with the new models and a new set of conformance scores as a new ViSQOL version.

**Comment by mchinen** [comment]
A fix for this was merged in #66.
